src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets

```
@@ -121,8 +121,93 @@ Copyright (c) .NET Foundation. All rights reserved.
     ============================================================
     -->
   <Target Name="CopyFilesToPublishDirectory"
-          DependsOnTargets="_CopyResolvedFilesToPublishPreserveNewest;
+          DependsOnTargets="_IncrementalCleanPublishDirectory;
+                            _CopyResolvedFilesToPublishPreserveNewest;
                             _CopyResolvedFilesToPublishAlways" />
+  
+  <!--
+    ============================================================
+                                        _IncrementalCleanPublishDirectory
+
+    Remove files that were produced in a prior build but weren't produced in the current build.
+    ============================================================
+    -->
+  <Target Name="_IncrementalCleanPublishDirectory"
+          DependsOnTargets="_GetCurrentAndPriorPublishFileWrites">
+
+    <!-- Subtract list of files produced in the prior publish from list of files produced in this publish. -->
+    <ItemGroup>
+      <_OrphanPublishFileWrites Include="@(_PriorPublishFileWrites)" Exclude="@(_CurrentPublishFileWrites)"/>
+    </ItemGroup>
+
+    <!-- Delete the orphaned files. -->
+    <Delete
+        Files="@(_OrphanPublishFileWrites)"
+        TreatErrorsAsWarnings="true">
+      <Output TaskParameter="DeletedFiles" ItemName="_OrphanFilesDeleted"/>
+    </Delete>
+    
+    <!-- Write new list of current files back to clean file. -->
+    <WriteLinesToFile
+        File="$(IntermediateOutputPath)$(PublishCleanFile)"
+        Lines="@(_CurrentPublishFileWrites)" 
+        Overwrite="true"/>  
+  </Target>
+
+  <!--
+    ============================================================
+                                        _GetCurrentAndPriorPublishFileWrites
+    Get the list of files written in the previous publish and the list of files to be written in this publish.
+    ============================================================
+    -->
+  <Target Name="_GetCurrentAndPriorPublishFileWrites" >
+    <PropertyGroup>
+      <NormalizedPublishDir>$([MSBuild]::NormalizeDirectory($(PublishDir)))</NormalizedPublishDir>
+    </PropertyGroup>
+    
+    <Hash ItemstoHash="$(NormalizedPublishDir)">
+      <Output TaskParameter="HashResult" PropertyName="NormalizedPublishDirHash" />
+    </Hash>
+    <PropertyGroup>
+      <PublishCleanFile Condition="'$(PublishCleanFile)'==''">.$(NormalizedPublishDirHash.Substring(0, 10)).txt</PublishCleanFile>
```

Should we make the filename more descriptive?
```suggestion
      <PublishCleanFile Condition="'$(PublishCleanFile)'==''">PublishOutputs.$(NormalizedPublishDirHash.Substring(0, 10)).txt</PublishCleanFile>
```